### ISSUE TITLE
Add flag to disable subsampling when data exceeds 1000 points

### ISSUE THREAD

**Issue by Thomas499** [schollz/find3#62]
The dashboard needs a little work. I have no idea how it is possible to have 0% accuracy on 56 readings.
![image](https://user-images.githubusercontent.com/38442058/38871093-d593166e-421d-11e8-90d8-6ffefc9e5dca.png)

Also, the number of data readings stored is always directly below the first character. This look great until you name a room something large. I.e. Beer pong room. Then the number of data readings directly overlap the accuracy readings.
![image](https://user-images.githubusercontent.com/38442058/38870999-816d20c0-421d-11e8-81a1-e50816616551.png)
![image](https://user-images.githubusercontent.com/38442058/38871011-8af27cda-421d-11e8-95c6-5005f2ebf962.png)

**Comment by Thomas499** [comment]
I am doing a few experiments using different families. The kttest family is one of the smaller data sets. The https://cloud.internalpositioning.com/view/dashboard/knewtek_test family is the experiment where I put the phone on a roomba to wonder each of the rooms. The knewtek_test family has plenty of data points (18971 currently). I mentioned the kttest family because that was the one that I could show the bug that I found.

**Comment by schollz** [comment]
@Thomas499 I'm not sure that this is a bug. I'll have to think about it, but realistically the accuracy of 0% should prompt the user to do more learning, which is necessary in the case here. Practically speaking this isn't usually going to be enough fingerprint data to actually use the system.

**Comment by Thomas499** [comment]
13 datapoints per location is plenty for the original find app to work. My biggest concern was that in all cases each location had at least 9 points of data to use in training and 3 points of data to use in testing. Out of 12 test data (3 data points each for testing * 4 locations) not one of them was correct. 

To put that into perspective, if you have multiple choice quiz where you fill in a, b, c, or d. Even if you randomly pick one you should get 25% correct. If you do this three times and don't get any correct, then even if you are picking randomly, statistically you should have more than 0% accuracy. 

No matter how many times I calibrate that data, which picks new training data and testing data each time, it is always 0%. 

I believe it is likely that this is a bug. - I will be happy to add more data points to if if you want me to. But then if it is a bug adding more data might hide the bug. - Let me know your thoughts.

**Comment by Thomas499** [comment]
Heres another example using a larger data set.

![image](https://user-images.githubusercontent.com/38442058/38931377-f7386fee-42e0-11e8-8e13-684a3c059997.png)
If there are 735 data points for second floor bathroom, then why is there also 0% accuracy for that location?

**Comment by schollz** [comment]
> Out of 12 test data (3 data points each for testing * 4 locations) not one of them was correct

This is not quite true. 0% does not mean incorrect nessecarily - the problem was that it couldn't actually predict *any* measurements because there were NaN values in parts of the confusion matrices. Its more precise to say that there is no information from your test data, because there is too little data.

Even though your accuracy is 0%, the learning could still work. Do you have any gauge on how well it works, even with these few data points? The accuracy is only an estimation of how it performs in the real-world, and is meant to give you an indicator of what to further strengthen with more learning. In the end, what matters is how well it performs in the real world (which is hard to measure unless you dedicate a lot of time).

Your second example is different, I believe. Could you share that dataset with me?

**Comment by Thomas499** [comment]
"In the end, what matters is how well it performs in the real world" - I agree.
"which is hard to measure unless you dedicate a lot of time" - I disagree. well depending on by what you mean by measure. It does not take a lot of time to make simple observations. 

Seriously, take your find3 app and go into a few rooms and see if the predictions are correct. My observations this morning are find3 is only 20-50% correct. 

I will send you a primitive apk that allows the original find and find3 to run within the same app if you want to use it to make observations on which is actually more accurate. 

"Your second example is different, I believe. Could you share that dataset with me?"
https://cloud.internalpositioning.com/view/dashboard/knewtek_short

**Comment by schollz** [comment]
@Thomas499 As I thought, the second scenario is indeed different.

The problem with your database is that it has *too many points that are not evenly distributed*. Specifically, the problem comes from these lines: https://github.com/schollz/find3/blob/c4179188acb406e70322dede4f44a4865b529661/server/main/src/api/calibration.go#L81-L83. Those lines basically truncate usesr who have data > 1000 points to only doing learning on a random subset of 1000 points. 

Why wouldn't I do learning on all of the data? Space is an issue. To compute the learning on your dataset it required 15 GB of RAM. However, after computing with all your data I can see that the accuracy improves greatly (88% -> 94%).

I assumed that most users will do the same amount of learning in each place. If so, then the random subsampling wouldn't matter because it randomly truncates points. However, your dataset has ~2000 datapoints in one location and <100 in another location, so the truncation disfavors the smaller locations. The solution to this is to either remove the truncation entirely, or do truncation only on specific locations.

Would it be useful to have a flag to disable this random subsampling (so you can use the whole dataset)? I will not be able to enable this on the public server because I can't afford that space to everyone (for free).

**Comment by Thomas499** [comment]
@schollz 

> I assumed that most users will do the same amount of learning in each place.
Bad idea, larger rooms will take longer and require more readings in different places of the room. Also, places that the user uses to charge the phone will also likely be subject to more readings.

I have two recommendations based on your previous response. 

1. If the 1000 limit is reached, start deleting older samples which are closest together based on the timestamps they were inserted. - remember what I said about doors opening and closing, and different humidity level. This method will be much more effective than simply capping the samples after first 1,000. Using this method, it really wouldn't even be necessary to include the flag. 1,000 samples should be plenty. The more spaced out they are the better. - It might also be beneficial to start the deletion selection from rooms that have vastly more readings than the others.


2. Why not evenly distribute the random truncate points from each location? i.e. make a rule if a location accounts for 15% of total readings or whatever, then 15% of the random truncate points must come from that location. - This would correct the mistake of assuming each location would have equal number of training data and would eliminate the disfavoring smaller locations issue. 

I am curious to find out more why 15GB of RAM is a concern of yours. For the learning process, I can understand that much may be needed, but that only happens when the person selects the calibrate machine learning option which does not happen often and will not happen after user finishes training. When the location identification code runs, this type of processing power will not be used (correct me if I am wrong). To me, it does not seem like the RAM issue should really be much of a concern. Now if you were using neural networking, where the a.i. constantly learns and corrects itself, I can see the processing requirement being more of an issue. 

What concerns me the most about the 15 GB of RAM statement you made is that android p is programmed to identify indoor location with accuracy of 1 to 2 meters, without the use of a server. I seriously doubt they would put that as a built in feature if it required 15GB of RAM on each phone.

**Comment by schollz** [comment]
1. Would be useful.
2. This is already what is happening.

Using a high amount of memory (15GB) is a concern for two reasons.

1. Personal reasons. I run a public server, for free. The learning is activated anytime there are new learning fingerprints inserted. The server is run on a computer that I use for other things. There are hundreds of users so there is learning happening all the time. Not every learning takes 15GB, but I don't want random people taking up all the RAM on my computer.
2. FIND3 can run on a Raspberry Pi, which is very useful for people who want to setup a cheap server.

Thus, a flag might be most useful. So if someone does have a system dedicated that they wish to use with thousands of datapoints, they can.

> What concerns me the most about the 15 GB of RAM statement you made is that android p is programmed to identify indoor location with accuracy of 1 to 2 meters, without the use of a server. I seriously doubt they would put that as a built in feature if it required 15GB of RAM on each phone.

You're right. Note that Android P has a entirely different system of internal positioning. They use specialized WiFi cards that have much higher time accuracy so they can do time-of-flight analysis to determine accurate distances to routers. These type of chips are rare in the consumer world, so I'm not sure how widely applicable Android P will be (hopefully very much so). Of course, they don't need 15GB of RAM, since they can directly measure the distances with a simple formula and then triangulate themselves.

**Comment by Thomas499** [comment]
> 2. This is already what is happening.

Your previous comments suggested otherwise. 

> I assumed that most users will do the same amount of learning in each place. If so, then the random subsampling wouldn't matter because it randomly truncates points. However, your dataset has ~2000 datapoints in one location and <100 in another location, so the truncation disfavors the smaller locations. 

> Personal reasons. I run a public server, for free. The learning is activated anytime there are new learning fingerprints inserted. The server is run on a computer that I use for other things. There are hundreds of users so there is learning happening all the time. Not every learning takes 15GB, but I don't want random people taking up all the RAM on my computer.

On the public server side, instead of the flag to disable the 1,000 cap only for private servers would you consider using a timestamp to act as the flag? The way I see it, you are on a 8 to 14 time hour difference based on the timing of your normal responses. You aren't going to use your computer between the hours of 11pm and 6am. Tweaking of hours as needed. I am not going to train during the hours of 11pm and 6am EST. This means I (and others in my region) will not likely be a noticeable drain on your RAM. 

Or cap the RAM usage allowed by the public server. That's the logical thing to do. 

By the way It is very nice of you to host this on your own computer as the server. However, that's probably why the find3 server is down right now.

**Comment by schollz** [comment]
@Thomas499 That's a good idea. I encourage you to fork the repo and add that as a pull request!

**Comment by schollz** [comment]
Closed because of https://github.com/schollz/find3/issues/66#issuecomment-382998197
